version: '3.7'

services:
  db:
    image: postgres
    restart: always
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
      - POSTGRES_DB=neuralshirts
    ports:
      - "5432:5432"
    volumes:
      - postgres:/var/lib/postgresql/data/
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "postgres"]
      interval: 1s
      timeout: 2s
      retries: 10

  api:
    build: .
    command: uvicorn autocoder.api.app:app --host 0.0.0.0 --port 8000
    ports:
      - "8000:8000"
    depends_on:
      - db
      - redis
      - celery
    healthcheck:
      test: curl --fail http://127.0.0.1:8000/healthcheck/ || exit 1 
      interval: 17s
      timeout: 5s
      retries: 3

  nginx:
    image: nginx:1.19
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - ./.config/nginx/default.conf:/etc/nginx/conf.d/default.conf
      - certbot-certs:/etc/letsencrypt
    depends_on:
      - api

  certbot:
     image: certbot/certbot
     volumes:
       - ./certbot/conf:/etc/letsencrypt
       - ./certbot/data:/var/lib/letsencrypt
     depends_on:
       - nginx
     command: certonly --webroot --webroot-path=/var/www/certbot --email parasol.torsion-0c@icloud.com --agree-tos --no-eff-email --force-renewal -d ${DOMAIN} -d www.${DOMAIN}

  celery:
    build: .
    command: celery -A main.celery_app worker --loglevel=info
    depends_on:
      - db
      - redis
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_RESULT_BACKEND=redis://redis:6379/0

  redis:
    image: redis:alpine
    ports:
      - "6379:6379"
    volumes:
      - redis:/data

volumes:
  postgres:
  redis:
  certbot-certs: